source/common/http/utility.cc

```diff
@@ -210,29 +292,45 @@ Utility::parseHttp1Settings(const envoy::api::v2::core::Http1ProtocolOptions& co
   return ret;
 }
 
-void Utility::sendLocalReply(StreamDecoderFilterCallbacks& callbacks, const bool& is_reset,
-                             Code response_code, const std::string& body_text) {
-  sendLocalReply(
-      [&](HeaderMapPtr&& headers, bool end_stream) -> void {
-        callbacks.encodeHeaders(std::move(headers), end_stream);
-      },
-      [&](Buffer::Instance& data, bool end_stream) -> void {
-        callbacks.encodeData(data, end_stream);
-      },
-      is_reset, response_code, body_text);
+void Utility::sendLocalReply(bool is_grpc, StreamDecoderFilterCallbacks& callbacks,
+                             const bool& is_reset, Code response_code,
+                             const std::string& body_text) {
+  sendLocalReply(is_grpc,
+                 [&](HeaderMapPtr&& headers, bool end_stream) -> void {
+                   callbacks.encodeHeaders(std::move(headers), end_stream);
+                 },
+                 [&](Buffer::Instance& data, bool end_stream) -> void {
+                   callbacks.encodeData(data, end_stream);
+                 },
+                 is_reset, response_code, body_text);
 }
 
 void Utility::sendLocalReply(
-    std::function<void(HeaderMapPtr&& headers, bool end_stream)> encode_headers,
+    bool is_grpc, std::function<void(HeaderMapPtr&& headers, bool end_stream)> encode_headers,
     std::function<void(Buffer::Instance& data, bool end_stream)> encode_data, const bool& is_reset,
     Code response_code, const std::string& body_text) {
+  // Respond with a gRPC trailers-only response if the request is gRPC
+  if (is_grpc) {
+    HeaderMapPtr response_headers{
+        new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Code::OK))},
+                          {Headers::get().ContentType, Headers::get().ContentTypeValues.Grpc},
+                          {Headers::get().GrpcStatus,
+                           std::to_string(enumToInt(httpToGrpcStatus(enumToInt(response_code))))}}};
+    if (!body_text.empty()) {
+      // TODO: GrpcMessage should be percent-encoded
+      response_headers->insertGrpcMessage().value(body_text);
+    }
+    encode_headers(std::move(response_headers), true); // Trailers only response
+    return;
+  }
+
   HeaderMapPtr response_headers{
       new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(response_code))}}};
   if (!body_text.empty()) {
     response_headers->insertContentLength().value(body_text.size());
     response_headers->insertContentType().value(Headers::get().ContentTypeValues.Text);
   }
-
+  // Somehow it's OK to encode headers even if 'is_reset' is true?
```

Comment:
What does this comment mean? The issue here is calling encode_headers() may result in a reset. You should be able to ASSERT(!is_reset); before this call. Can you add that and change the comment to be a bit better?

Reply:
I did not understand why the following `encode_data()` call is only done if `!is_reset`, but this preceding `encode_headers()` is not conditional on `is_reset` at all. Also, if ASSERT before this is proper, then the conditional on `!is_reset` before the following `encode_data()` is not needed any more, and the whole purpose of the `is_reset` parameter is for us to ASSERT on it?

Reply:
It's because the stream can get reset recursively as a result of the encodeHeaders() call. I realize this is probably not intuitive, but that's why this whole dance exists. Can you update comments and add relevant asserts? I think it will make it more clear.

Reply:
Done, took a while for me realize that `is_reset` can change before the encode_data() call as it is a reference. I'll add a comment to clarify that as well.